pkg/hub/taints/controller.go

```diff
@@ -0,0 +1,166 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	listerv1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	v1 "open-cluster-management.io/api/cluster/v1"
+	"time"
+)
+
+const (
+	NoCondition      = "NoManagedClusterConditionAvailable"
+	ConditionUnknown = "ManagedClusterConditionAvailableUnknown"
+	ConditionFalse   = "ManagedClusterConditionAvailableFalse"
+)
+
+// taintsController
+type taintsController struct {
+	kubeClient    kubernetes.Interface
+	clusterClient clientset.Interface
+	clusterLister listerv1.ManagedClusterLister
+	eventRecorder events.Recorder
+}
+
+// NewTaintsController creates a new taints controller
+func NewTaintsController(
+	kubeClient kubernetes.Interface,
+	clusterClient clientset.Interface,
+	clusterInformer informerv1.ManagedClusterInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &taintsController{
+		kubeClient:    kubeClient,
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("taints-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("taintsController", recorder)
+}
+
+func (c *taintsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling Taints %s", managedClusterName)
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	managedCluster = managedCluster.DeepCopy()
+	newTaints := managedCluster.Spec.Taints
+
+	switch {
+	case meta.IsStatusConditionTrue(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		newTaints, _ = c.deleteTaintAndJudgeExist(newTaints, "",
+			v1.ManagedClusterTaintUnreachable, v1.ManagedClusterTaintUnavailable)
+	case meta.IsStatusConditionFalse(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		var exist bool
+		newTaints, exist = c.deleteTaintAndJudgeExist(newTaints,
+			v1.ManagedClusterTaintUnavailable, v1.ManagedClusterTaintUnreachable)
+		if !exist {
+			t1 := v1.Taint{
+				Key:    v1.ManagedClusterTaintUnavailable,
+				Value:  ConditionFalse,
+				Effect: v1.TaintEffectNoSelect,
+				TimeAdded: metav1.Time{
+					Time: time.Now(),
+				},
+			}
+			newTaints = append(newTaints, t1)
+		}
+	case c.isIsStatusConditionUnknown(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) ||
+		meta.FindStatusCondition(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) == nil:
```

Comment:
this may need another case to switch

Reply:
and do we really need to judge this?

Reply:
It can be split into two cases, but the logic of the two cases is very similar, only the value is different.